### PR TITLE
Improve abort handling

### LINKED ISSUE #31404: Abort gracefully closes the response rather than aborting it
```
Assert.DoesNotContain() Failure
Found:    (filter expression)
In value: ConcurrentQueue<WriteContext> [WriteContext { EventId = 0, Exception = null, Formatter = Func3 { ... }, LoggerName = "Microsoft.AspNetCore.Server.IIS.FunctionalTests.Te"..., LogLevel = Debug, ... }, WriteContext { EventId = Starting, Exception = null, Formatter = Func3 { ... }, LoggerName = "Microsoft.Extensions.Hosting.Internal.Host", LogLevel = Debug, ... }, WriteContext { EventId = HostingStartupAssemblyLoaded, Exception = null, Formatter = Func3 { ... }, LoggerName = "Microsoft.AspNetCore.Hosting.Diagnostics", LogLevel = Debug, ... }, WriteContext { EventId = 0, Exception = null, Formatter = Func3 { ... }, LoggerName = "Microsoft.Hosting.Lifetime", LogLevel = Information, ... }, WriteContext { EventId = 0, Exception = null, Formatter = Func`3 { ... }, LoggerName = "Microsoft.Hosting.Lifetime", LogLevel = Information, ... }, ...]
```

```
at Microsoft.AspNetCore.Server.IIS.FunctionalTests.StrictTestServerTests.Dispose() in /_/src/Servers/IIS/IIS/test/IIS.Tests/StrictTestServerTests.cs:line 17
   at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource) in C:\Dev\xunit\xunit\src\xunit.execution\Extensions\ReflectionAbstractionExtensions.cs:line 79
```

<details>
<summary>Logs</summary>

```
[0.001s] [TestLifetime] [Information] Starting test ClosesAfterDataSent at 2021-03-30T06:36:31
[6.836s] [Microsoft.AspNetCore.Server.IIS.FunctionalTests.TestServer] [Debug] Method: GET, RequestUri: 'http://localhost:58030/start', Version: 1.1, Content: <null>, Headers:
{
}
[6.839s] [Microsoft.Extensions.Hosting.Internal.Host] [Debug] Hosting starting
[6.839s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Debug] Loaded hosting startup assembly IIS.Tests
[6.840s] [Microsoft.Hosting.Lifetime] [Information] Application started. Press Ctrl+C to shut down.
[6.840s] [Microsoft.Hosting.Lifetime] [Information] Hosting environment: Production
[6.840s] [Microsoft.Hosting.Lifetime] [Information] Content root path: C:\h\w\B64C0933\w\AE29096B\e\.
[6.840s] [Microsoft.Extensions.Hosting.Internal.Host] [Debug] Hosting started
[6.840s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request starting HTTP/1.1 GET http://localhost:58030/start - -
[6.841s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request finished HTTP/1.1 GET http://localhost:58030/start - - - 200 - - 0.3827ms
[6.841s] [Microsoft.AspNetCore.Server.IIS.FunctionalTests.TestServer] [Debug] StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Server: Microsoft-IIS/10.0
  Date: Tue, 30 Mar 2021 06:36:37 GMT
  Content-Length: 4
}
[6.841s] [Microsoft.AspNetCore.Server.IIS.FunctionalTests.TestServer] [Debug] Done
[6.843s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request starting HTTP/1.1 POST http://localhost/ - 1
[6.966s] [Microsoft.Hosting.Lifetime] [Information] Application is shutting down...
[7.523s] [Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer] [Error] Connection ID "18086456105130524677", Request ID "80000006-0000-fb00-b63f-84710c7967bb": An unhandled exception was thrown by the application.
System.Runtime.InteropServices.COMException (0x80070006): The handle is invalid. (0x80070006 (E_HANDLE))
   at Microsoft.AspNetCore.Server.IIS.NativeMethods.Validate(Int32 hr) in /_/src/Servers/IIS/IIS/src/NativeMethods.cs:line 331
   at Microsoft.AspNetCore.Server.IIS.NativeMethods.HttpTryCancelIO(NativeSafeHandle pInProcessHandler) in /_/src/Servers/IIS/IIS/src/NativeMethods.cs:line 280
   at Microsoft.AspNetCore.Server.IIS.Core.IO.AsyncIOEngine.Complete() in /_/src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs:line 148
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync() in /_/src/Servers/IIS/IIS/src/Core/IISHttpContextOfT.cs:line 95
[7.523s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request finished HTTP/1.1 POST http://localhost/ - 1 - 200 - - 680.6435ms
[9.058s] [TestLifetime] [Information] Finished test ClosesAfterDataSent in 9.0567293s
```

</details>

https://dev.azure.com/dnceng/public/_build/results?buildId=1062917&view=ms.vss-test-web.build-test-results-tab&runId=32739106&resultId=110110&paneView=debug

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/dotnet/aspnetcore/issues/31404

### CODE REVIEW

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs [hunk 1]
```diff
@@ -159,6 +159,11 @@ private async Task WriteBody(bool flush = false)
         {
             while (true)
             {
+                if (_bodyOutput.Aborted)
+                {
+                    break;
```
Comment (HaoK) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs:164]: Not really sure this break here is ok, as I'm not really sure what the expectations are before/after this loop, is this what you had in mind @Tratcher ?

Comment (HaoK) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs:164]: We have tests explicitly checking that Writes after abort no-op, so I don't think we want to throw exceptions right?

File: src/Servers/IIS/IIS/src/Core/OutputProducer.cs [hunk 2]
```diff
@@ -9,9 +9,10 @@ namespace Microsoft.AspNetCore.Server.IIS.Core;
 
 internal class OutputProducer
 {
-    // This locks access to _completed.
+    // This locks access to _completed and _aborted.
     private readonly object _contextLock = new object();
     private bool _completed;
+    private bool _aborted;
```
Comment (Tratcher) [inline src/Servers/IIS/IIS/src/Core/OutputProducer.cs:15]: volatile? You set it in a lock, but read it outside.

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs (hunk 1 above)
Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs:164]: Right, Write shouldn't throw.

I think this check needs to be after ReadAsync or you'd miss an Abort that happened while waiting for data.

Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs:166]: Hmm, there might be an issue if AdvanceTo isn't called? Moving this check inside the Try block would cover that with the finally.

**Review by Tratcher — APPROVED** [review]

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs (hunk 1 above)
Comment (davidfowl) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs]: Are we always calling Complete on the pipe? If we don't there's a memory leak.

**Comment by davidfowl** [comment]
> Hmm, there might be an issue if AdvanceTo isn't called? Moving this check inside the Try block would cover that with the finally.

You can skip AdvanceTo if you call Complete. Complete must always be called.

**Comment by HaoK** [comment]
The complete call is in the finally outside of the while loop, so I think we are good there.  So should we just break at the top?  I think the only question is where this new check for aborted needs to go

**Comment by Tratcher** [comment]
I think the current version is fine. The important part is that the abort check is after ReadAsync.

**Pull request merged**